### PR TITLE
Improve JavaScript behavior at coordinator startup

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Improve JavaScript behavior at coordinator startup
Forward port of https://github.com/arangodb/arangodb/pull/12770
Documentation PR: https://github.com/arangodb/docs/pull/563
Oskar PR: https://github.com/arangodb/oskar/pull/270

- [x] :hankey: Bugfix 
- [x] :pizza: New feature 
- [x] :hammer: Refactoring 
- [x] :book: CHANGELOG entry made
- [x] :muscle: The behavior in this PR was *manually tested*
- [ ] :computer: The behavior change can be verified via automatic tests

#### Backports:

- [x] Backports required for: 3.6, 3.7

### Testing & Verification

http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/12113/

### CODE REVIEW

**Comment by dothebart** [comment]
should there be a way to stop one coordinator from being (or continuing to be) a foxx-master?
If server is started with foxx-queues disabled, should it check whether its foxx-master and turn that down in concequence so a new leader is elected?

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/12322/

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/12330/

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/12345/

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/12347/

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/12354/

File: Documentation/DocuBlocks/foxxStartup.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+if *true*, all Foxx services in all databases will be synchronized between
+multiple coordinators during the boot sequence. This ensures that all Foxx
+services are up-to-date when a coordinator reports itself as ready.
+
+In case the option is set to `false` (i.e. no waiting), the coordinator 
+will complete the boot sequence faster, and the Foxx services will be 
+propagated lazily. Until the initialization procedure has completed for
+the local Foxx apps, any request to a Foxx app will be responded to with
+an HTTP 500 error and message
```
Comment (mchacki) [inline Documentation/DocuBlocks/foxxStartup.md:14]:
```suggestion
an HTTP 503 error and message 
```

**Review by mchacki — CHANGES_REQUESTED** [review]
All we have here is a good PR and worth merging.

However I would like to see a combined test of
`test-restart-cluster` and `foxx`.

e.g.
1) install A Foxx App.
2) Suspend DBServers
3) Start a new Coordinator. (this should start, but cannot sync up Foxx Apps)
4) Request Foxx on this Coordinator => Service Unavailable.
5) Unsuspend DBServers
6) Retry accessing Foxx => Eventually the service will get available.

I am happy to have this kind of test in resilience tests suite as an alternative.

File: Documentation/DocuBlocks/foxxStartup.md (hunk 1 above)
Comment (jsteemann) [reply Documentation/DocuBlocks/foxxStartup.md:14]: It actually returns HTTP 500, even though in my code I used 503. This is due to some exception rewriting inside Foxx.

File: CHANGELOG [hunk 2]
```diff
@@ -1,6 +1,41 @@
...
+
+  This can cause an unavailability window for Foxx services on coordinator
+  startup for the initial requests to Foxx apps until the app propagation 
+  has completed.
+  
+  When not using Foxx, this option should be set to `false` (default) to 
+  benefit from a faster coordinator startup.
+  Deployments relying on Foxx apps being available as soon as a coordinator 
+  is integrated or responding should set this option to `true`.
+  The option only has an effect for setups with multiple coordinators.
```
[the remark this answers is not in the document]
Comment (jsteemann) [reply CHANGELOG]: I think I already had the wording you are suggesting in my PR, but somehow Github doesn't always show the latest state.

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/12376/

**Review by mchacki — APPROVED** [review]
LGTM.
Thanks for adding the test :+1:

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/12391/

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/12402/

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/12420/

**Pull request merged**